sdk/storage/Azure.Storage.Files.DataLake/src/Models/GetFileSystemsAsyncCollection.cs

```diff
@@ -28,26 +28,165 @@ internal class GetFileSystemsAsyncCollection : StorageCollectionEnumerator<FileS
+            /// An async sequence of <see cref="Page{T}"/>s.
+            /// </returns>
+            public override IEnumerable<Page<FileSystemItem>> AsPages(
+                string continuationToken = default,
+                int? pageHintSize = default)
+            {
+                return _enumerator._client.GetBlobContainers(
+                    (BlobContainerTraits)_enumerator._traits,
+                    _enumerator._prefix,
+                    CancellationToken)
+                    .AsPages(continuationToken, pageHintSize)
+                    .Select(ConvertPage);
+            }
+
+            /// <summary>
+            /// Enumerate the values in the collection synchronously.  This may
+            /// make mutliple service requests.
+            /// </summary>
+            /// <returns>A sequence of values.</returns>
+            public override IEnumerator<FileSystemItem> GetEnumerator()
+            {
+                return _enumerator._client.GetBlobContainers(
+                    (BlobContainerTraits)_enumerator._traits,
+                    _enumerator._prefix,
+                    CancellationToken)
+                    .Select(ConvertItem)
+                    .GetEnumerator();
+            }
+        }
+
+        /// <summary>
+        /// Abstract the Storage pattern for async iteration
+        /// </summary>
+        private class StorageAsyncPageable : AsyncPageable<FileSystemItem>
+        {
+            private GetFileSystemsAsyncCollection _enumerator;
+
+            // for mocking
+            protected StorageAsyncPageable()
+                : base()
+            {
+            }
+
+            public StorageAsyncPageable(GetFileSystemsAsyncCollection enumerator, CancellationToken cancellationToken)
+                : base(cancellationToken)
+            {
+                _enumerator = enumerator;
+            }
+
+            /// <summary>
+            /// Enumerate the values a <see cref="Page{T}"/> at a time.  This may
+            /// make mutliple service requests.
+            /// </summary>
+            /// <param name="continuationToken">
+            /// A continuation token indicating where to resume paging or null to
+            /// begin paging from the beginning.
+            /// </param>
+            /// <param name="pageHintSize">
+            /// The size of <see cref="Page{T}"/>s that should be requested (from
+            /// service operations that support it).
+            /// </param>
+            /// <returns>
+            /// An async sequence of <see cref="Page{T}"/>s.
+            /// </returns>
+            public override async IAsyncEnumerable<Page<FileSystemItem>> AsPages(
+                string continuationToken = default,
+                int? pageHintSize = default)
+            {
+                await foreach (Page<BlobContainerItem> page in
+                    _enumerator._client.GetBlobContainersAsync(
+                        (BlobContainerTraits)_enumerator._traits,
```

Comment:
Nit - might be nice to pull this out into another `Convert` method because we've got a few of these